docs/decisions/0003-saga-service-testing.rst

```diff
@@ -0,0 +1,186 @@
+
+Decision
+--------
+
+These tests use redux-saga's "runSaga" method where possible to assert the interaction between a saga, its invoking action, and any side effects (primarily service methods) that it may in turn invoke.
+
+Sagas which delegate to other sagas may still be tested via method #2 above, as running them using runSaga doesn't otherwise produce any output.
+
+Method #1 (primary test method)
+===============================
+
+This method is effective, but requires quite a bit of setup and mocking.
+
+Most sagas will call a service method that makes API calls against a server.  The API client our services use to make requests can be mocked, which allows us to test the contract between the saga and service.
+
+To mock the API client's methods, configure the service with a mock apiClient object where any AJAX methods (post, get, delete, patch, etc.) are implemented with Jest mocks.
+
+::
+
+  // Example body of the server response.
+  const responseData = {
+    data: {
+      // Data pertinent to this specific example - yours will be different.
+      voucher: { id: 12345, code: 'DEMO25', benefit: { type: PERCENTAGE_BENEFIT, value: 25 },
+      },
+    },
+  };
+
+  const apiClientPost = jest.fn(() =>
+    new Promise((resolve, reject) => {
+      resolve(responseData); // resolving here results in a successful server response.
+    }));
+
+  // If you aren't using our standard service configuration methods, then this code may vary for you.  The crux is that you want to give a mock apiClient to the service.
+  configureApiService(
+    { ECOMMERCE_BASE_URL: 'http://localhost' }, // Any necessary "configuration" the service needs.
+    { post: apiClientPost }, // Mocked apiClient methods
+  );
+
+Next, the saga under test may be run using runSaga:
+
+::
+
+  it('should handle success', async () => {
+
+    // ... The above mocked apiClient and configured service...
+
+    const dispatched = []; // Used to record any actions the saga will dispatch
+    await runSaga(
+      { dispatch: action => dispatched.push(action) },
+      handleAddCoupon, // The saga under test
+      addCoupon('DEMO25'), // The action to feed to the saga
+    ).toPromise(); // Important!  Along with the await and async test method above, allows us to wait for the saga to finish.
+
+Finally, we can assert the saga and service have run as expected:
+
+::
+
+  // Asserting that the actions we expect were dispatched in the correct order, with the correct payloads.
+  expect(dispatched).toEqual([
+    addCouponBegin(),
```

Comment:
I think I did comment on this in an earlier PR. Should we be comparing a raw object instead of the return value of the action creator?